edx_notifications/stores/store.py

```diff
@@ -0,0 +1,152 @@
+    set in NOTIFICATION_STORE_PROVIDER in the settings file
+
+    NOTE: When we switch over to gevent support, we should use some
+    locking techniques in this area to prevent concurrent execution
+    """
+
+    global _STORE_PROVIDER  # pylint: disable=global-statement
+
+    if not _STORE_PROVIDER:
+        config = getattr(settings, 'NOTIFICATION_STORE_PROVIDER')
+        if not config:
+            raise ImproperlyConfigured("Settings not configured with NOTIFICATION_STORE_PROVIDER!")
+
+        if 'class' not in config or 'options' not in config:
+            msg = (
+                "Misconfigured NOTIFICATION_STORE_PROVIDER settings, "
+                "must have both 'class' and 'options' keys."
+            )
+            raise ImproperlyConfigured(msg)
+
+        module_path, _, name = config['class'].rpartition('.')
+        class_ = getattr(import_module(module_path), name)
+
+        _STORE_PROVIDER = class_(**config['options'])
+
+    return _STORE_PROVIDER
+
+
+def reset_notification_store():
+    """
+    Tears down any cached configuration. This is useful for testing.
+
+    NOTE: When we switch over to gevent support, we should use some
+    locking techniques in this area to prevent concurrent execution
+    """
+
+    global _STORE_PROVIDER  # pylint: disable=global-statement
+
+    _STORE_PROVIDER = None
+
+
+class BaseNotificationStoreProvider(object):
+    """
+    The base abstract class for all notifications data providers, such as MySQL/Django-ORM backed.
+
+    IMPORTANT: NotificationStoreProvider is assumed to be a singleton, therefore there must be
+    no state stored in the instance of the provider class.
+    """
+
+    # don't allow instantiation of this class, it must be subclassed
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def get_notification_message_by_id(self, msg_id, options=None):
+        """
+        Returns the notitication message (of NotificationMessage type) by primary key
+
+        ARGS:
+            - msg_id: the primary key of the NotificationMessage
+            - options: dictionary of options. Possible choices:
+                * 'select_related': whether to fully fetch any related objects
+
+        RETURNS: type NotificationMessage
+        """
+
+    @abc.abstractmethod
+    def save_notification_message(self, msg):
+        """
+        Save (create or update) a notification message (of NotificationMessage type)
+
+        ARGS:
+            - msg: an instance of NotificationMessage. If the 'id' field is
+                   set by the caller, then it is assumed to be an update
+                   operation
+
+        RETURNS: type NotificationMessage
+        """
+
+    def save_notification_user_map(self, user_map):
+        """
+        Create or Update the mapping of a user to a notification.
+        """
+
+    @abc.abstractmethod
+    def get_notification_type(self, name):
+        """
+        This returns a NotificationType object.
+        """
```

Comment:
From #3: It would be good to raise NotImplementError in these methods, so that if someone super chains up to them, they'll actually fail, rather than just returning None.

I tend to follow the guidelines that abstract methods that change state should have empty base methods (no state change), and that abstract methods that return things should raise errors (no way to return an unknown empty thing from the base). And I try not to have methods that do both. =)